File: 0402-remove-k-digits/0402-remove-k-digits.py
# TC- O(n)
# SC- O(n)

class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        
        # goal is to move smaller elements as fas as possible to the left
        stack = [] # monotonically inc

        for ch in num:
            digit = int(ch)
            while k and stack and digit<int(stack[-1]):
                stack.pop()
                k -= 1
            stack.append(ch)
        
        if k:
            stack = stack[:-k] # if k elements weren't removed yet, then remove elements from tail
        
        return "".join(stack).lstrip('0') or "0" # remove leading 0s

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
// TC- O(n)
// SC- O(n)

class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<char> st;
        for(int i=0; i<num.size(); i++){
            int curr = num[i];
            while(!st.empty() and st.top()>curr and k>0){
                st.pop();
                k--;
            }
            st.push(curr);
        }
        while(k>0 and !st.empty()){
            st.pop();
            k--;
        }
        
        string res;
        while(!st.empty()){
            res += st.top();
            st.pop();
        }   
        reverse(res.begin(), res.end());
        // cout<<res;

        int i=0;
        while(i<res.size() and res[i]=='0'){
            i++;
        }
        res = res.substr(i);
        return (res.empty())? "0":res;
    }
};